mountpoint-s3/src/data_cache/express_data_cache.rs

```diff
@@ -149,16 +149,19 @@ where
         // Guarantee that the request will start even in case of `initial_read_window == 0`.
         result.as_mut().increment_read_window(self.config.block_size as usize);
 
-        // TODO: optimize for the common case of a single chunk.
-        let mut buffer = BytesMut::default();
+        let mut buffer: Option<BytesMut> = None;
         while let Some(chunk) = result.next().await {
             match chunk {
                 Ok((offset, body)) => {
-                    if offset != buffer.len() as u64 {
+                    if offset != buffer.as_ref().map_or(0, |buffer| buffer.len()) as u64 {
                         emit_failure_metric_read("invalid_block_offset");
                         return Err(DataCacheError::InvalidBlockOffset);
                     }
-                    buffer.extend_from_slice(&body);
+                    if let Some(ref mut buffer) = buffer {
+                        buffer.extend_from_slice(&body);
+                    } else {
+                        buffer.replace(BytesMut::from(Bytes::from(body)));
+                    }
```

Comment:
I'd suggest to have `buffer: Bytes` and then just do:
```suggestion
                    buffer = if buffer.is_empty() {
                        Bytes::from(body)
                    } else {
                        // Unlikely: we expect `get_object` to return a single chunk.
                        let mut buffer = BytesMut::from(buffer);
                        buffer.extend_from_slice(&body);
                        buffer.freeze()
                    };
```

It would also get rid of the `map`s elsewhere.